crates/storage/src/transactional.rs

```diff
@@ -1,96 +1,811 @@
 //! The primitives to work with storage in transactional mode.
 
-use crate::Result as StorageResult;
+use crate::{
+    kv_store::{
+        BatchOperations,
+        KeyValueInspect,
+        KeyValueMutate,
+        StorageColumn,
+        Value,
+        WriteOperation,
+    },
+    structured_storage::StructuredStorage,
+    Result as StorageResult,
+};
+use std::{
+    collections::{
+        btree_map::Entry,
+        BTreeMap,
+        HashMap,
+    },
+    sync::Arc,
+};
 
-#[cfg_attr(feature = "test-helpers", mockall::automock(type Storage = crate::test_helpers::EmptyStorage;))]
-/// The types is transactional and may create `StorageTransaction`.
-pub trait Transactional {
-    /// The storage used when creating the transaction.
-    type Storage: ?Sized;
-    /// Creates and returns the storage transaction.
-    fn transaction(&self) -> StorageTransaction<Self::Storage>;
+#[cfg(feature = "test-helpers")]
+use crate::{
+    iter::{
+        BoxedIter,
+        IterDirection,
+        IterableStore,
+    },
+    kv_store::KVItem,
+};
+
+/// Provides a view of the storage at the given height.
+/// It guarantees to be atomic, meaning the view is immutable to outside modifications.
+pub trait AtomicView: Send + Sync {
+    /// The type of the storage view.
+    type View;
+
+    /// The type used by the storage to track the commitments at a specific height.
+    type Height;
+
+    /// Returns the latest block height.
+    fn latest_height(&self) -> Option<Self::Height>;
+
+    /// Returns the view of the storage at the given `height`.
+    fn view_at(&self, height: &Self::Height) -> StorageResult<Self::View>;
+
+    /// Returns the view of the storage for the latest block height.
+    fn latest_view(&self) -> Self::View;
+}
+
+/// Storage transaction on top of the storage.
+pub type StorageTransaction<S> = StructuredStorage<StorageTransactionInner<S>>;
+
+/// The inner representation of the storage transaction.
+#[doc(hidden)]
```

Comment:
`#[doc(hidden)]` should be explained in a comment here. Is this type within semver guarantees? Why it shouldn't be used outside this crate? Why is it available?